### ISSUE TITLE
Docs: Update DNS forwarding Guide for `systemd-resolved`

### ISSUE THREAD

**Issue by mkeeler** [hashicorp/consul#5985]
Currently our [`systemd-resolved` section of the dns forwarding guide](https://learn.hashicorp.com/consul/security-networking/forwarding#systemd-resolved-setup) mentions that `systemd-resolved` requires the upstream DNS server to be bound to port 53 and we show how you can have Consul bound to port 8600 and use `iptables` rules to redirect port 8600 to 53 to make the integration work.

That integration only works if the DNS server in Consul is listening on the loopback interface. The `iptables` man page explains it pretty clearly:

> REDIRECT
> 
> This target is only valid in the nat table, in the PREROUTING and OUTPUT chains, and user-defined chains which are only called from those chains. It redirects the packet to the machine itself by changing the destination IP to the primary address of the incoming interface (locally-generated packets are mapped to the 127.0.0.1 address). It takes one option:
>
> –to-ports port [- port ]
>
>This specifies a destination port or range of ports to use: without this, the destination port is never altered. This is only valid if the rule also specifies -p tcp or -p udp .

Basically it will auto-map the IP to 127.0.0.1 when you use a redirect rule. Therefore we should update the guide to mention this restriction.

Either you bind Consul to port 53 or Consul must be serving DNS on the 127.0.0.1 address. If neither of those two requirements are met then `systemd-resolved` DNS forwarding to Consul will not work on a single machine/vm/container.

You could still forward DNS requests from one machine to another on a different port using `iptables` DNAT rules instead of REDIRECT rules. Whether we should mention that in the guide I am not certain of. Thats getting pretty complex.

**Comment by rberlind** [comment]
I found that I had to add the following to my Consul configuration to make this work:
```
recursors = [
  "10.64.0.254"
]
```
where `10.64.0.254 was the DNS server I wanted DNS queries to domains other than .consul forwarded to.

I also had to restart the `systemd-resolved` service with this:
```
systemctl restart systemd-resolved
```

Also, I actually added the systemd-resolved config to a new file, /etc/systemd/resolved.conf.d/consul.conf, instead of to /etc/systemd/resolved.conf, although the latter probably would have also worked.

**Comment by imcom** [comment]
@rberlind  I now quite believe that is is a `systemd-resolved` issue here, it does not respect the so called `route-only` domains and DNS servers configuration. It simply uses the first available DNS server to resolve all domains given to it ... So `~consul` on `127.0.0.1` for example won't be used.

**Comment by imcom** [comment]
```
Global
         DNS Servers: 172.31.0.2
                      127.0.0.1
          DNS Domain: ~consul
                      ap-southeast-1.compute.internal
```
Here is my configuration for `systemd-resolved`
```
[Resolve]
DNS=127.0.0.1
Domains=~consul
```
If I do not have a global DNS setting, this will intercept all DNS queries to Consul, this is why @rberlind  needed to set a recursor to Consul. But once I have some other Global setting, like `/etc/systemd/resolved.conf` with content:
```
[Resolve]
DNS=172.31.0.2
```
Then Consul won't get the chance to resolve `.consul.` domains

**Comment by rberlind** [comment]
Interesting.

**Comment by Wolfsrudel** [comment]
I have the same problem. How can we solve this?

**Comment by mkeeler** [comment]
I was able to get things working like:

```
$> consul agent -dev -client 10.1.0.32
```

```
[Resolve]
DNS=10.1.0.32
Domains=~consul
```

```
$> iptables -t nat -A OUTPUT -p tcp -d 10.1.0.32 --dport 53 -j DNAT --to-destination 10.1.0.32:8600 
$> iptables -t nat -A OUTPUT -p udp -d 10.1.0.32 --dport 53 -j DNAT --to-destination 10.1.0.32:8600 
$> iptables -t nat -A PREROUTING -p tcp -d 10.1.0.32 --dport 53 -j DNAT --to-destination 10.1.0.32:8600 
$> iptables -t nat -A PREROUTING -p udp -d 10.1.0.32 --dport 53 -j DNAT --to-destination 10.1.0.32:8600
```

For those that need other Global resolvers to be configured I think the best course of action would be to configure those as recursors in the Consul configuration and omit the `Domains=~consul` setting. This will send *all* DNS queries to Consul and Consul will use the configured recursors to handle domains it is not authoritative for.

**Comment by mrlesmithjr** [comment]
Was about to start digging into this myself rather than just dumping `systemd-resolve` like I normally do. What is the latest progress on this? Or are the Consul docs on this topic functional?

**Comment by akhenakh** [comment]
Future release of systemd-resolved 246 will support pointing to a different port:
```
A space-separated list of IPv4 and IPv6 addresses to use as system DNS servers. Each address can optionally take a port number separated with ":",
```
https://www.freedesktop.org/software/systemd/man/resolved.conf.html

**Comment by destructobeam** [comment]
Just to confirm the above, the following (now) works.

```
# /etc/systemd/resolved.conf.d/consul.conf
[Resolve]
DNS=127.0.0.1:8600
Domains=~.consul
```

**Comment by blake** [comment]
> Future release of systemd-resolved 246 will support pointing to a different port:

The [DNS forwarding](https://learn.hashicorp.com/tutorials/consul/dns-forwarding) tutorial was updated yesterday to add instructions for configuring systemd-resolved to forward Consul queries directly to port 8600 when using systemd 246 or newer.

I'll work to get a note added that the documented iptables workaround only functions if Consul's DNS is listening on the loopback address.

**Comment by blake** [comment]
A note was just added after the iptables instructions which states:

> The above configuration assumes Consul's DNS server is listening on the loopback address. If Consul is not listening on the loopback IP, replace the references to 'localhost' and '127.0.0.1' with the appropriate IP address for your environment.

I've closed this issue because I believe this change addresses the feedback that was initially reported in this issue, and yesterday's changes address the other comments around systemd-resolved's configuration.

Please let me know if this is not the case, and I will either re-open this issue, or create a new issue to track subsequent feedback.

**Comment by mrlesmithjr** [comment]
Thanks for the info @blake
